master/buildbot/test/unit/reporters/test_generators_buildrequest.py

```diff
@@ -0,0 +1,156 @@
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+
+from parameterized import parameterized
+
+from mock import Mock
+
+from twisted.internet import defer
+from twisted.trial import unittest
+
+from buildbot.reporters.generators.buildrequest import BuildRequestGenerator
+from buildbot.test.fake import fakemaster
+from buildbot.test.util.config import ConfigErrorsMixin
+from buildbot.test.util.misc import TestReactorMixin
+from buildbot.test.util.reporter import ReporterTestMixin
+
+
+class TestBuildRequestGenerator(ConfigErrorsMixin, TestReactorMixin,
+                                unittest.TestCase, ReporterTestMixin):
+
+    all_messages = ('failing', 'passing', 'warnings', 'exception', 'cancelled')
+
+    def setUp(self):
+        self.setUpTestReactor()
+        self.setup_reporter_test()
+        self.master = fakemaster.make_master(self, wantData=True, wantDb=True,
+                                             wantMq=True)
+
+    @parameterized.expand([
+        ('tags', 'tag'),
+        ('tags', 1),
+        ('builders', 'builder'),
+        ('builders', 1),
+        ('schedulers', 'scheduler'),
+        ('schedulers', 1),
+        ('branches', 'branch'),
+        ('branches', 1),
+    ])
+    def test_list_params_check_raises(self, arg_name, arg_value):
+        kwargs = {arg_name: arg_value}
+        g = BuildRequestGenerator(**kwargs)
+        with self.assertRaisesConfigError('must be a list or None'):
+            g.check()
+
+    def setup_generator(self, message=None, **kwargs):
+        if message is None:
+            message = {
+                "body": "start body",
+                "type": "plain",
+                "subject": "start subject"
+            }
+
+        g = BuildRequestGenerator(**kwargs)
+
+        g.formatter = Mock(spec=g.formatter)
+        g.formatter.format_message_for_build.return_value = message
+
+        return g
+
+    @defer.inlineCallbacks
+    def test_build_message_start_no_result(self):
+        g = yield self.setup_generator()
+        buildrequest = yield self.insert_buildrequest_new()
+        build = yield g.partial_build_dict(self.master, buildrequest)
+        report = yield g.buildrequest_message(self.master, build)
+
+        g.formatter.format_message_for_build.assert_called_with(self.master, build,
+                                                                mode=self.all_messages,
+                                                                users=[])
+
+        self.assertEqual(report, {
+            'body': 'start body',
+            'subject': 'start subject',
+            'type': 'plain',
+            'results': None,
+            'builds': [build],
+            'users': [],
+            'patches': [],
+            'logs': []
+        })
+
+    @defer.inlineCallbacks
+    def test_is_message_needed_ignores_unspecified_tags(self):
```

Comment:
This test is no longer needed as the is_message_needed_by_props is tested in test_generators_utils. I see that TestBuildStartEndGenerator has it too so existing code has the same problem.

I think we could change this test to call `generate` like the test_generate_new below. This way we'll cover the case when generate does not generate a report. The test would need to be moved to the end of the file in that case.